src/Microsoft.Extensions.WebSockets/WebSocketFrame.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Text;
+using Channels;
+
+namespace Microsoft.Extensions.WebSockets
+{
+    /// <summary>
+    /// Represents a single Frame received or sent on a <see cref="IWebSocketConnection"/>.
+    /// </summary>
+    public struct WebSocketFrame
+    {
+        /// <summary>
+        /// Indicates if the "FIN" flag is set on this frame, which indicates it is the final frame of a message.
+        /// </summary>
+        public bool EndOfMessage { get; }
+
+        /// <summary>
+        /// Gets the <see cref="WebSocketOpcode"/> value describing the opcode of the WebSocket frame.
+        /// </summary>
+        public WebSocketOpcode Opcode { get; }
+
+        /// <summary>
+        /// Gets the payload of the WebSocket frame.
+        /// </summary>
+        public ReadableBuffer Payload { get; }
+
+        public WebSocketFrame(bool endOfMessage, WebSocketOpcode opcode, byte[] payload) : this(endOfMessage, opcode, new ArraySegment<byte>(payload)) { }
+
+        public WebSocketFrame(bool endOfMessage, WebSocketOpcode opcode, ArraySegment<byte> payload) : this(endOfMessage, opcode, ReadableBuffer.Create(payload.Array, payload.Offset, payload.Count)) { }
```

Comment:
This is *not* a cheap way to create buffers; IMO the ideal would be to keep the data "as is" until you can actually write it to the *actual* buffer. The way I do this in my version is a bit cheeky, but works great:

    var buffer = out.Alloc();
    WriteFrameHeader(whatever);
    buffer.Write(theOriginalValue);
    if(mask != 0) {
        var asReadable = buffer.AsReadableBuffer(); // slice and dice whatever so we're just looking at the payload
        ApplyMask(asReadable, mask);
    }
    await buffer.FlushAsync();

This also has the advantage that you only need to be able to write masking code on readable buffers... which sounds a little anti-intuitive, but it works fine.